### ISSUE TITLE
[Keras Fit & Compile] Horovod worker desynchronization - Potentially deadlock training TF 2.4

### ISSUE THREAD

**Issue by DEKHTIARJonathan** [tensorflow/tensorflow#46863]
CC: @reedwm @pkanwar23 @sanjoy 
CC2: @romerojosh @tgaddair 
CC3: @nluehr @WhiteFangBuck

We unfortunately found quite a serious bug in TF2.4.X with Keras Models (Compile and Fit approach) combined with Mixed Precision Policy when using Horovod.

Models can ran into Deadlock or DeSynchronization (all reduce gradient of one step from worker A with gradient of another step of worker B) and stop training completely.

To make the matter worse, this issue increases in likelihood with the number of workers.

I have put together a simple repro case showing how workers can deadlock and basically stalls:

https://github.com/DEKHTIARJonathan/HVD_TF_Worker_Desync_BugReport

Small extract of the bug:
```bash
[2021-02-02 21:19:07.345988: W /tmp/pip-install-dxtho7yu/horovod_0bde588442554cf4b80c76441efc4e03/horovod/common/stall_inspector.cc:105] One or more tensors were submitted to be reduced, gathered or broadcasted by subset of ranks and are waiting for remainder of ranks for more than 10 seconds. This may indicate that different ranks are trying to submit different tensors or that only subset of ranks is submitting tensors, which will cause deadlock. 
Missing ranks:
0: [cond_1/then/_10/cond_1/SGD/PartitionedCall/DistributedSGD_Allreduce/cond/then/_79/DistributedSGD_Allreduce/cond/HorovodAllreduce_grads_0]
```
As a note, this bug wasn't present in prior versions of TF.

Any help would be greatly appreciated.

Thanks a lot

**Comment by romerojosh** [comment]
I wanted to provide a bit more detail on this issue and the behavior we are observing. 

We use the [Horovod](https://github.com/horovod/horovod) library to run distributed training. Horovod operates by modifying the `get_gradients` method of the optimizer to return an globally averaged (allreduced) gradient in place of local gradients (https://github.com/horovod/horovod/blob/2a775b2e059817093e476a5a44433426dda35ae5/horovod/_keras/__init__.py#L71-L81) It seems that there is currently an incompatibility with the Kera `LossScaleOptimizer` in TF 2.4.x that is causing issues with this functionality related to the loss scaling/step skipping functionality.

In the repro case posted by @DEKHTIARJonathan, we have provided a small reproducer script that builds a trivial model in Keras and runs a single training step, with the mixed precision policy enabled or disabled via arg `--use_amp`. In the script, we set the input of rank 0 to a NaN value to generate a NaN gradient, and the input of rank 1 to a sane input value to generate a non-NaN gradient. 

The expected behavior of this test is that during the call to `get_gradients`, the gradient will be allreduced by Horovod and produce a common NaN across the two workers. Then in `apply_gradients`, the `LossScaleOptimizer` will detect this on both workers and they both skip the optimizer step. 

However, this is not the behavior we are seeing. Instead, we find that the worker that produces the NaN gradient (rank 0), skips the step independently of worker 1. Even worse, the worker that skips the step appears to completely bypass launching any Horovod operations at all, as though it skips the call to `get_gradients` entirely. The result of this is the output shown in the posted repo:
```
1/1 - 3s - loss: nan
rank 0 completed step...


[2021-02-02 21:19:07.345988: W /tmp/pip-install-dxtho7yu/horovod_0bde588442554cf4b80c76441efc4e03/horovod/common/stall_inspector.cc:105] One or more tensors were submitted to be reduced, gathered or broadcasted by subset of ranks and are waiting for remainder of ranks for more than 10 seconds. This may indicate that different ranks are trying to submit different tensors or that only subset of ranks is submitting tensors, which will cause deadlock. 
Missing ranks:
0: [cond_1/then/_10/cond_1/SGD/PartitionedCall/DistributedSGD_Allreduce/cond/then/_79/DistributedSGD_Allreduce/cond/HorovodAllreduce_grads_0]
```
In this output, we see that rank 0 is able to return from the `model.fit()` call while rank 1 is blocked on the call. The Horovod stall message printed indicates that the reason rank 1 is stalled is because rank 0 never launches its matching Horovod allreduce op for the gradient listed. 

We see that this stall is remedied by using TF 2.3.x or if we disable the mixed precision policy in Keras using TF 2.4.x. 

@reedwm There was a long thread a while back working out Horovod interactions with Keras in #36398. The Horovod Keras implementation hasn't changed that much since then. Perhaps you are aware of something that changed in TF 2.4.x that might explain this new behavior?

**Comment by ravikyram** [comment]
@DEKHTIARJonathan 

Can you please share colab link or simple standalone code to reproduce the issue in our environment.It helps us in localizing the issue faster. Thanks!

**Comment by DEKHTIARJonathan** [comment]
@ravikyram that won't be doable I'm sorry. This issue is specifically a multi-GPU issue. So AFAIK, there's no way to use Google Colab to do that.

And you don't need to use the docker container I provided if you prefer to build everything yourself on bare metal.
